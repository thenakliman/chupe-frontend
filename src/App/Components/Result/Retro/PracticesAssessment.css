.practices {
  width: 65rem;
  margin: 0 auto;
}

.practices__heading {
  font-size: 1.4rem;
  padding: 1rem 0.5rem;
  background: #dcd0c0;
}

.description {
  background: #fff;
  font-size: 1.2rem;
  display: flex;
  justify-content: space-between;
  padding: 1rem 0.2rem;
  border-bottom: 1px solid #ccc;
}

.description:hover {
  background-color: #f9f9f9;
  transition: background-color 0.5s ease;
}

.description__btn {
  width: min-content;
  transform: scale(2);
  margin: 0 1rem;
}

.description__label {
  margin-right: 1rem;
}