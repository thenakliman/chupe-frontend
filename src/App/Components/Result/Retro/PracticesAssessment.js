import React from 'react';
import propTypes from 'prop-types';
import {history} from '../../../utils/history';

require('./PracticesAssessment.css');

export class PracticesAssessment extends React.Component {
  constructor(props) {
    super(props);
    this.state = {assessments: []};
    this.onPracticeAssessmentChange = this.onPracticeAssessmentChange.bind(this);
    this.doesFollow = this.doesFollow.bind(this);
    this.savePracticesAssessment = this.savePracticesAssessment.bind(this);
  }

  componentWillMount() {
    this.props.fetchPractices();
  }

  savePracticesAssessment() {
    this.props.savePracticeAssessment(this.props.match.params.id, this.state.assessments)
        .then(() => history.push(`/retro/${this.props.match.params.id}`));
  }

  onPracticeAssessmentChange(id, answer) {
    const assessmentPosition = this.state.assessments.findIndex((assessment) => assessment.bestPracticeId === id);
    const assessments = [...this.state.assessments];
    assessments[assessmentPosition].answer = answer;
    this.setState({assessments: assessments});
  }

  doesFollow(id) {
    let assessments = this.state.assessments;
    if (assessments.length === 0) {
      assessments = this.props.practices.map(
          (assessment) => ({bestPracticeId: assessment.id, answer: true}));
      this.setState({assessments: assessments});
    }

    return assessments.find((assessment) => assessment.bestPracticeId === id).answer;
  }

  render() {
    return <div className={'practices'}>
      <div className={'practices__heading'}>
        Best Practices
      </div>
      {this.props.practices.map((practice, index) => (
          <div key={index} className='description'>
            <div className={'description__text'}>{practice.description}</div>
            <div className={'description__btns'}>
              <label className={'description__label'}>
                <input className={'description__btn'}
                       id={`yes-${index}`}
                       type={'checkbox'}
                       name={'applicable'}
                       onChange={() => this.onPracticeAssessmentChange(practice.id, true)}
                       checked={this.doesFollow(practice.id)}/>
                Yes
              </label>
              <label className={'description__label'}>
                <input className={'description__btn'}
                       id={`no-${index}`}
                       type={'checkbox'}
                       name={'applicable'}
                       onChange={() => this.onPracticeAssessmentChange(practice.id, false)}
                       checked={!this.doesFollow(practice.id)}/>
                No
              </label>
            </div>
          </div>
      ))}
      <button className={'button--small u-margin-top-small'}
              onClick={this.savePracticesAssessment}>
        Save
      </button>
    </div>;
  }
}

PracticesAssessment.propTypes = {
  practices: propTypes.array.isRequired,
  fetchPractices: propTypes.func.isRequired,
  savePracticeAssessment: propTypes.func.isRequired,
};
