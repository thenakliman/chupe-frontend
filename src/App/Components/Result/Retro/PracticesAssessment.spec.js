/* eslint-disable */
import React from 'react';
/* eslint-enable */
import {mount, shallow} from 'enzyme';
import {PracticesAssessment} from "./PracticesAssessment";
import toJson from 'enzyme-to-json';

describe('practices assessments', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<PracticesAssessment fetchPractices={jest.fn()}
                                                 savePracticeAssessment={jest.fn()}
                                                 practices={[{id: 1010, description: "some practices"}]}/>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should fetch assessments on mount', () => {
    const fetchPractices = jest.fn();
    shallow(<PracticesAssessment fetchPractices={fetchPractices}
                                 savePracticeAssessment={jest.fn()}
                                 practices={[{id: 1010, description: "some practices"}]}/>);

    expect(fetchPractices).toHaveBeenCalledWith();
  });

  it('should change state as per answer for a practice', async () => {
    const fetchPractices = jest.fn();
    const practices = [{
      id: 1010, description: "some practices - 1"
    }, {
      id: 1012, description: "some practices - 2"
    }];

    const wrapper = mount(<PracticesAssessment fetchPractices={fetchPractices}
                                               savePracticeAssessment={jest.fn()}
                                               practices={practices}/>);
    await wrapper.find('#no-0').simulate('change');

    expect(wrapper.state().assessments).toEqual([{bestPracticeId: 1010, answer: false}, {
      bestPracticeId: 1012,
      answer: true
    }]);
  });

  it('should save assessment on click of save', () => {
    const savePracticeAssessment = jest.fn(() => Promise.resolve());
    const practices = [{
      id: 1010, description: "some practices - 1"
    }, {
      id: 1012, description: "some practices - 2"
    }];

    const params = {params: {id: 1}};
    const wrapper = shallow(<PracticesAssessment fetchPractices={jest.fn()}
                                                 savePracticeAssessment={savePracticeAssessment}
                                                 match={params}
                                                 practices={practices}/>);
    const assessments = [{bestPracticeId: 1010, answer: false}, {bestPracticeId: 1012, answer: true}];
    wrapper.setState({assessments: assessments});

    wrapper.find('button').simulate('click');
    expect(savePracticeAssessment).toHaveBeenCalledWith(1, assessments);
  });
});