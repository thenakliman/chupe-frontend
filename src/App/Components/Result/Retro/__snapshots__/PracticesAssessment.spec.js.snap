// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`practices assessments should match snapshot 1`] = `
<div
  className="practices"
>
  <div
    className="practices__heading"
  >
    Best Practices
  </div>
  <div
    className="description"
    key="0"
  >
    <div
      className="description__text"
    >
      some practices
    </div>
    <div
      className="description__btns"
    >
      <label
        className="description__label"
      >
        <input
          checked={true}
          className="description__btn"
          id="yes-0"
          name="applicable"
          onChange={[Function]}
          type="checkbox"
        />
        Yes
      </label>
      <label
        className="description__label"
      >
        <input
          checked={false}
          className="description__btn"
          id="no-0"
          name="applicable"
          onChange={[Function]}
          type="checkbox"
        />
        No
      </label>
    </div>
  </div>
  <button
    className="button--small u-margin-top-small"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
